File: some-python/networking/network-sniffer.py
import socket
import struct

def format_mac(bytes_addr):
	return ":".join(f'{b:02x}' for b in bytes_addr)

def format_ip(bytes_addr):
	return '.'.join(str(b) for b in bytes_addr)

def main() :
	raw_socket = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.htons(0x0003))

	print("Listening for raw ethernet frames...")

	try:
		while True:
			raw_data, _ = raw_socket.recvfrom(65535)
			dest_mac, src_mac, eth_proto = struct.unpack('! 6s 6s H', raw_data[:14])
			
			print(f"\n[Ethernet Frame]")
			print(f" Source MAC: {format_mac(src_mac)} -> Dest MAC: {format_mac(dest_mac)} | Protocol: {hex(eth_proto)}")

			if eth_proto == 0x0800:
				ip_header = raw_data[14:34]
				version_ihl, tos, total_length, pkt_id, flags_offset, ttl, proto, checksum, src_ip, dest_ip = struct.unpack('! B B H H H B B H 4s 4s', ip_header)
				ihl = (version_ihl & 0x0F) *4

				print(f" [IPv4 Packet]")
				print(f" Source IP: {format_ip(src_ip)} -> Dest IP: {format_ip(dest_ip)} | Protocol: {proto} | TTL: {ttl}")
				
				if proto == 6:
					tcp_start = 14 + ihl
					tcp_header = raw_data[tcp_start:tcp_start + 20]
					
					src_port, dest_port, seq, ack, offset_reserved_flags, flags = struct.unpack('! H H L L B B', tcp_heade[:14])
					flag_syn = (flags & 0x02) >> 1
					flag_ack = (flags & 0x10) >> 4
					flag_fin = (flags & 0x01)
					print(f"[TCP Segment]")
					print(f"Source Port: {src_port} -> Dest Port: {dest_port}")
					print(f"Sequence: {seq} | Ack: {ack}")
					print(f"Flags -> SYN: {flag_syn} | ACK: {flag_ack} | FIN: {flag_fin}")

	except KeyboardInterrupt:
			print("\n Stopping sniffer")
			raw_socket.close()

if __name__ == "__main__":
	main()

File: some-c/networking/network-sniffer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <net/ethernet.h>
#include <netinet/if_ether.h>
#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <sys/socket.h>

#define BUFFER_SIZE 65536

void process_packet(unsigned char *buffer, int size)
{
	struct ethhdr *eth = (struct ethhdr *)buffer;
	
	printf("\n[Ethernet Frame]\n");
	printf("Source MAC: %02x:%02x:%02x:%02x:%02x:%02x -> Dest MAC: %02x:%02x:%02x:%02x:%02x:%02x | Protocol: 0x%4x\n", eth->h_source[0], eth->h_source[1], eth->h_source[2], eth->h_source[3],eth->h_source[4],eth->h_source[5],eth->h_dest[0],eth->h_dest[1],eth->h_dest[2],eth->h_dest[3],eth->h_dest[4],eth->h_dest[5],ntohs(eth->h_proto));

	if (ntohs(eth->h_proto) == ETH_P_IP)
	{
		struct iphdr *ip = (struct iphdr *)(buffer + sizeof(struct ethhdr));

		struct in_addr src_addr, dest_addr;
		src_addr.s_addr = ip->saddr;
		dest_addr.s_addr = ip->daddr;

		printf("[IPv4 Packet]\n");
		printf("Source IP: %s -> Dest IP: %s | Protocol: %d | TTL: %d\n", inet_ntoa(src_addr), inet_ntoa(dest_addr), ip->protocol, ip->ttl);

		if (ip->protocol == IPPROTO_TCP)
		{
			int ip_header_len = ip->ihl*4;
			
			struct tcphdr *tcp = (struct tcphdr *)(buffer + sizeof(struct ethhdr) + ip_header_len);
			printf("[TCP Segment]\n");
			printf("Source Port: %d -> Dest Port: %d\n", ntohs(tcp->source), ntohs(tcp->dest));
			printf("Sequence: %u | ACK: %u\n", ntohl(tcp->seq), ntohl(tcp->ack_seq));
			printf("Flags -> SYN: %d | ACK: %d | FIN %d\n", tcp->syn, tcp->ack, tcp->fin);
		}
	}
}

int main()
{
	int raw_sock;
	unsigned char buffer[BUFFER_SIZE];

	raw_sock = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
	if (raw_sock < 0)
	{
		perror("Socket createion error");
		return 1;
	}

	printf("Listening for raw ethernet frames... \n");

	while (1)
	{
		ssize_t data_size = recvfrom(raw_sock, buffer, BUFFER_SIZE, 0, NULL, NULL);
		if (data_size < 0)
		{
			perror("Recv Error");
			return 1;
		}
		
		process_packet(buffer, data_size);
	}

	close(raw_sock);
	return 0;
}
